runtime/vm/jfr.cpp

```diff
@@ -635,17 +655,130 @@ tearDownJFR(J9JavaVM *vm)
 	(*vmHooks)->J9HookUnregister(vmHooks, J9HOOK_VM_INITIALIZED, jfrVMInitialized, NULL);
 
 	/* Free global data */
+	freeJFRConstantEvents(vm);
+
 	j9mem_free_memory((void*)vm->jfrBuffer.bufferStart);
 	memset(&vm->jfrBuffer, 0, sizeof(vm->jfrBuffer));
 	if (NULL != vm->jfrBufferMutex) {
 		omrthread_monitor_destroy(vm->jfrBufferMutex);
 		vm->jfrBufferMutex = NULL;
 	}
+	if (NULL != vm->jfrState.isConstantEventsInitializedMutex) {
+		omrthread_monitor_destroy(vm->jfrState.isConstantEventsInitializedMutex);
+		vm->jfrState.isConstantEventsInitializedMutex = NULL;
+	}
 	j9mem_free_memory(vm->jfrState.metaDataBlobFile);
 	vm->jfrState.metaDataBlobFile = NULL;
 	vm->jfrState.metaDataBlobFileSize = 0;
 }
 
+/**
+ * Helper to get the value of a system property.
+ *
+ * @param vm[in] the J9JavaVM
+ * @param propName[in] the system property name
+ */
+static const char *getSystemProp(J9JavaVM *vm, const char *propName) {
+	J9VMSystemProperty *jvmInfoProperty = NULL;
+	const char *value = "";
+	UDATA getPropertyResult = vm->internalVMFunctions->getSystemProperty(vm, propName, &jvmInfoProperty);
+	if (J9SYSPROP_ERROR_NOT_FOUND != getPropertyResult) {
+		value = jvmInfoProperty->value;
+	}
+	return value;
+}
+
+/**
+ * Initialize constantEvents.
+ *
+ * @param vm[in] the J9JavaVM
+ */
+static void
+initializeJFRConstantEvents(J9JavaVM *vm)
+{
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	/* Allocate constantEvents */
+	vm->jfrState.constantEvents = j9mem_allocate_memory(sizeof(JFRConstantEvents), J9MEM_CATEGORY_VM);
+
+	initializeJVMInformationEvent(vm);
+}
+
+/**
+ * Free constantEvents.
+ *
+ * @param vm[in] the J9JavaVM
+ */
+static void
+freeJFRConstantEvents(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	freeJVMInformationEvent(vm);
+
+	j9mem_free_memory(vm->jfrState.constantEvents);
+}
+
+/**
+ * Initialize JVMInformationEntry.
+ *
+ * @param vm[in] the J9JavaVM
+ */
+static void initializeJVMInformationEvent(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	/* Initialize JVM Information */
+	JVMInformationEntry *jvmInformation = &(((JFRConstantEvents *)(vm->jfrState.constantEvents))->JVMInfoEntry);
+
+	/* Set JVM name */
+	jvmInformation->jvmName = getSystemProp(vm, "java.vm.name");
+
+	/* Set JVM version */
+	jvmInformation->jvmVersion = getSystemProp(vm, "java.vm.info");
+
+	/* Set Java arguments */
+	jvmInformation->javaArguments = getSystemProp(vm, "sun.java.command");
+
+	/* Set JVM arguments by concatenating actualVMArgs */
+	JavaVMInitArgs *vmArgs = vm->vmArgsArray->actualVMArgs;
+	UDATA vmArgsLen = vmArgs->nOptions - 1;
```

Comment:
why is this initialized to the number of args?

Reply:
You can just do `UDATA vmArgsLen = vmArgs->nOptions` and not add +1. for the null terminator.